### ISSUE TITLE
Auto confirmation does not work if server is not startet or screen is locked.

### ISSUE THREAD

**Issue by inktrap** [IzzySoft/Adebar#14]
- If I do not start the server and show the connected devices with `àdb devices`, the auto-confirmation will fail.
- this also happens when the screen is locked.
- Basically everything that delays the backup-operation will result in the right keypress being send in the wrong situation.
- I am trying to fix this, but to cover all eventualities it would be necessary to ensure that the backup-dialog/screen is the currently active window and that nothing else (like a keyboard) is present. This is not possible via adb.
- it should at least be documented that the device needs to be connected first and the server has to be running and the screen should be unlocked.

Just keeping this here for future reference.

**Comment by IzzySoft** [comment]
Item 1 should be obvious – at least the part with the started server :) I've noticed, though, that the first call seems to always fail – while all subsequent calls work fine (tested on my O4X when it still had the stock 4.1.1 Android this afternoon; now it's running 4.4 (CM11), and I will try again as soon as time permits, should be this week).

Item 2 [is already documented](https://github.com/IzzySoft/Adebar/wiki/Known-Issues#backup-of-each-app-has-to-be-confirmed-separately) and also obvious if you think about it: Security measure to prevent "unauthorized backups" (though those should already be prevented by "auth").

Number 3 I thought to be a show-stopper (or guess why I had to build-in delays already, and even an option to explicitly confirm the start of the next restore from the command line? ADB got a hickup on some devices when a second restore came in while the first was still in progress). Amazingly, your key-event trigger worked like a charm: I've been sitting with eyes wide open, seeing a completely automated backup of each app into a single archive!

Shall I add item 4 next to were item 2 is documented – or rather wait if you might figure something? I'd address the "common sense" that one shouldn't play with the device while the backup process is in progress :)

**Comment by inktrap** [comment]
I guess trying to solve these issues would require some kind of app.
Controlling the keyboard is not possible via adb, I am relatively sure
about this.

I think that all these things are obvious, but should be mentioned in
the docs. Therefore bugs regarding this can be easily closed/won't
occur. So, yes, add item 4 to item 2.

Glad you liked my changes :)

Small excursion about device security:
- Unlocking the screen (if no password is set) should be possible
  `adb shell input keyevent 82` (did not test this)
- You can get a shell if debugging is enabled and the device is locked
- You can disable the set password via adb + root access if you change the
  sqlite-db, c.f.:
  http://forum.xda-developers.com/showthread.php?t=1409304 (note, I
  did not try this, but it should work like this or similarly)
- so yes, if you enable debugging, your device is open to an attack from
  an already authorized device, f.e. there is malware/PUA that tries
  to do this.

**Comment by IzzySoft** [comment]
> So, yes, add item 4 to item 2.

I was just about to do that – but then I wondered: as soon as you trigger an `adb backup` or `adb restore\* command, the backup-dialog/screen automatically becomes the active/foreground screen. So what should the user ensure there (except for not touching the screen while the "automatic process" is running)? So maybe we should instead print "RUNNING AUTOMATED PROCESS – DO NOT TOUCH YOUR DEVICE!!!" in bold, uppercase, red, blinking letters on the shell runing _Adebar?_ :)

> Glad you liked my changes

_Liked?_ That's an understatement, actually :)

> Unlocking the screen (if no password is set) should be possible `adb shell input keyevent 82` (did not test this)

Negative here. Tried with the screen off (it didn't even switch on), and with the "screen lit". Nothing happened (2 devices running CM11 / Android 4.4, "swipe to unlock"). But then tried a 3rd device (tablet running stock 4.1), and it worked. Doesn't do any harm in either case; so what about an `AUTO_UNLOCK` option? If set to `1`, the keyevent would be transmitted right before the first backup/restore command, and give it a 1s delay to register? The documentation then could mention this doesn't work with all devices/ROMs, but trying it out doesn't do any damage.

And yes, dangers increase with _USB debugging_ enabled and the device rooted (guess that's why we can define whether ADB might obtain root). It increases even more with ADB running in "insecure mode" (so a simple `adb shell` would give you a root-shell).

**Comment by inktrap** [comment]
> So what should the user ensure there (except for not touching the screen while the "automatic process" is running)? So maybe we should instead print "RUNNING AUTOMATED PROCESS – DO NOT TOUCH YOUR DEVICE!!!" in bold, uppercase, red, blinking letters on the shell runing Adebar? :)

I would not make it that annoying, something plain like should do the job:

```
NOTICE: Close all open windows and leave your device unlocked!
NOTICE: Do not interact with your device while performing backups or restores!
```

I would not clutter the Adbar-config with an option (AUTO_UNLOCK) that should be the responsiblity of the user and which may not work reliably. This would just give users the impression that Adbar takes care of this stuff and frees them from the responsibility to follow the NOTICE above. Adbar can't and won't do that.

So yup, If I have more ideas/improvements I will tell you. :)

**Comment by IzzySoft** [comment]
> NOTICE: Close all open windows and leave your device unlocked!

I don't feel this one necessary:
- irritation: where to close them – PC or device? 
- usefulness: why to close them? I've never done so, and never experienced any issues. The backup dialog comes to front anyway.

> NOTICE: Do not interact with your device while performing backups or restores!

This sounds reasonable. I'd throw that with `AUTO_CONFIRM` being enabled (done with b97d664).

> I would not clutter the Adbar-config with an option (AUTO_UNLOCK) that should be the responsiblity of the user and which may not work reliably.

Too late – see 19b7337. But it's turned off by default.

> This would just give users the impression that Adbar takes care of this stuff and frees them from the responsibility to follow the NOTICE above. Adbar can't and won't do that.

I hope I've pointed that out sufficiently in the comments and the wiki (the later not yet visible publically):

```
# Shall Adebar try to automatically unlock your device when running your
# backup/restore script? Doesn't work with all devices/ROMs, and requires
# that no password/pin/pattern is set.
AUTO_UNLOCK=1
```

> - `AUTO_UNLOCK`: Shall the backup/restore scripts try to automatically unlock your screen? Note that this a) doesn't work on all devices, b) on some devices might require the screen already being turned on, and c) definitely doesn't work if a password/pattern/pin protection is active. Hence by default this is turned off. Feel free to play with it: in the worst case, it simply doesn't work.

Again, thanks a lot for your contribution! This definitely gave _Adebar_ a kick up. Having to confirm each app backup/restore manually was the one (and hopefully only) big annoyance.

**Comment by inktrap** [comment]
Excerpts from Izzy's message of 2015-11-14 15:07:12 +0100:

> > NOTICE: Close all open windows and leave your device unlocked!
> 
> I don't feel this one necessary:
> - irritation: where to close them – PC or device? 
> - usefulness: why to close them? I've never done so, and never experienced any issues. The backup dialog comes to front anyway.

Yes, that is correct (I have not thought about that ambiguity). Well, if
I personally back something up, I make sure that I am up to the task,
that includes no distractions, full battery aso. and I close all running
stuff. But that is just my personal preference.

> > This would just give users the impression that Adbar takes care of this stuff and frees them from the responsibility to follow the NOTICE above. Adbar can't and won't do that.
> 
> I hope I've pointed that out sufficiently in the comments and the wiki (the later not yet visible publically):

Yup, looks nice. But I assume that most users do not read the docs.
Because it is off by default, you have to look into the config anyway
and the warning there should be alright.

> Again, thanks a lot for your contribution! This definitely gave _Adebar_ a kick up. Having to confirm each app backup/restore manually was the one (and hopefully only) big annoyance.

My pleasure!
So far, I have found nothing else :clap:

**Comment by IzzySoft** [comment]
> Yup, looks nice. But I assume that most users do not read the docs. Because it is off by default, you have to look into the config anyway and the warning there should be alright.

That was my intention. Not propagating it too much, having it turned off by default, users _might_ find it "by accident" and wonder what it's for – so they have to read at least the comments. Hopefully they'd even check with the wiki for further hints. I couldn't get it to work reliably with most of my devices – and though that's not really representative, I've decided to rather "hide" it. Maybe it rises someone's interest, and we get a little more feedback on it. This worked well with some MTK stuff in the past (which I made throwing some warnings when triggered as I have no MTK device to test – so as soon as someone used it on any, it'd turn up): my "debug output" turned up, the user created an issue here – et voila, I had my tester :)

> So far, I have found nothing else :clap:

That's a good thing (in terms that everything "works as advertized"). Still, ideas are always welcome :smiley_cat:
